File: Data Structures & Algorithms/count-good-nodes-in-binary-tree/submission-0.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

class Solution:
    def goodNodes(self, root: TreeNode) -> int:
        count=0
        dq=deque([[root,root.val]])
        while dq:
            for i in range(len(dq)):
                list=dq.popleft()
                node=list[0]
                if node.val==list[1]:
                    count+=1
                if node.left:
                    dq.append([node.left,max(node.left.val,list[1])])
                if node.right:
                    dq.append([node.right,max(node.right.val,list[1])])
        return count

File: Data Structures & Algorithms/kth-smallest-integer-in-bst/submission-0.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
#include <vector>

class Solution {
public:
    int kthSmallest(TreeNode* root, int k) {
        int leftCount = count(root->left)+1;
        if (leftCount==k) {
            return root->val;
        } if (leftCount>k) {
            return kthSmallest(root->left,k);
        }  return kthSmallest(root->right,k-leftCount);
    }

    int count(TreeNode* root) {
        if (root==nullptr) {
            return 0;
        } return 1 + count(root->left) + count(root->right);
    }
};
